ldap/servers/slapd/mapping_tree.c

```diff
@@ -569,7 +569,7 @@ free_mapping_tree_node_arrays(backend ***be_list, char ***be_names, int **be_sta
  * tree node (guaranteed to be non-NULL).
  */
 static int
-mapping_tree_entry_add(Slapi_Entry *entry, mapping_tree_node **newnodep)
+mapping_tree_entry_add(Slapi_Entry *entry, mapping_tree_node **newnodep, PRBool check_be)
```

Comment:
Is there a situation where we would want check_be to be false?

Reply:
Yes, when we start the start the server we load the entries from dse.ldif out of order, so we skip the check in this case otherwise the server fails to start :-)

Reply:
As in we load the MT entries out of order? Or we load the MT before the BE's?

Reply:
Correct, the mapping tree entry is processed before the backend entry.  So technically you can still screw things up if you edit the dse.ldif manually.  But we only support doing these updates over LDAP so I'm not that worried about it to be honest.

Reply:
Really? For the mapping tree node to exist it has to have the BE pointer that it's going to serve, so the BE has to be loaded first. That's what get_backend_from_attr relies on in the entry_add call.

Reply:
If it didn't happen I would not have had to add the boolean check :-)

Reply:
Oh I believe you but I'm confused. We rely on the backend list being populated at start up, else we'll get the error "The nsslapd-backend attribute has no value for the mapping tree", so the backend list *has* to be populated first, else no backends could possibly be loaded in the mapping tree without a subsequent second pass over the configs. So I'm curious now to see how the situation exists where the PR_FALSE flag is needed here because in my mind it just doesn't compute, and probably indicates some other issue ....

Reply:
Oh I just found the comment in get_backends_from_attr that says "we'll fix it at runtime". Yeesh. I wonder if there is a cleaner or better way to handle this situation then, because it would be good to have these checks able to always run to assert the relevant backend does exist and matches.

Reply:
So in the startup case, we indeed init the mapping trees before starting the backend plugin.  The default backends (config, schema, rootdse) are inited, but not the ldbm backends:

main() --> setup_internal_backends() -->  mapping_tree_init() --> mapping_tree_node_get_children() --> mapping_tree_entry_add()

Reply:
All good, thanks for clearing that up. In that case Ack from me :)